### ISSUE TITLE
.net version upgrade

### ISSUE THREAD

**Issue by jan-guenter** [chocolatey/choco#2436]
### TL;DR
I was surprised to find that chocolatey still uses .net Framework 4, so I took a quick look on how complicated an upgrade to .net 5 would be and created a proof of concept.
The goal of this issue is to gauge if there is even an interest in upgrading and to gather feedback before committing more time or opening a PR.

---

### Background
I planned to improve a little bit on the chocolatey GUI and was surprised to see the project using .net Framework 4, so I decided to start from scratch using .net 5 and WPF to create a more native feeling UI.
When trying to add the chocolatey lib I realized that it also is only available for .net Framework.
So I decided to have look how complicated it would be to upgrade choco.
To my surprise the upgrade was rather painless and made a POC in my fork.
There would still a lot of things left to due to finalize.
Now I have my POC in a rather decent shape I think and want to if there is any desire to upgrade and to get feedback before committing more time to it.

---

### Pros and Cons of upgrading
The only reason I can imagine for sticking with .net Framework 4 is support for Windows Vista which ended with 4.6.
I don't know how many choco users still use Vista and I deprecating Vista support is an option at all, so would be glad to get some input on this.

In my opinion, the most prominent reason to upgrade is that .net Framework 4 end of life was in 2016.
There is a multitude of reasons for upgrading which should be fairly obvious, so I won't bore you with too much details, but for the sake of completeness I will list some major bullet points:
- ability to use newer libraries
- replacing hacks and workarounds (e.g. TLS support, filesystem path length issue, etc.)
- new language features improving code quality, readability and maintainability
- new core features replacing custom implementations

---

### Proof Of Concept
You can find the proof of concept code `dotnet-5` branches in my forks of `choco` and `nuget-chocolatey`:
https://github.com/jan-guenter/choco/tree/dotnet-5
https://github.com/jan-guenter/nuget-chocolatey/tree/dotnet-5

#### **choco**
I upgraded the project in two steps; first to .net Framework 4.8 and then to .net 5.
In it's current state the project is builds and runs fine. All the unit tests succeed and only 3 of the integration test scenarios fail. The in all three failing scenarios only the `should_not_be_able_delete_the_rollback` test fails, which I was not able to fix on the first glance.

Here is a list of the changes I made _(omitting some smaller adjustment details)_:
- upgraded projects to .net 5
- upgraded dependencies to latest versions
- removed `AlphaFS` dependency _(all features are available in .net 5 including support for long paths)_
- replaced Rx-* packages with `System.Reactive`
- replaced `Should` library with `FluentAssertions`
- replaced local version of `System.Management.Automation` with latest upstream nuget package
- added `Microsoft.PowerShell.SDK` and `Microsoft.PowerShell.Commands.Diagnostics` to complement `System.Management.Automation`
- added local .net 5 version of the chocolatey `NuGet.Core` library
- adjusted to breaking changes in SimpleInjector
- adjusted tests to breaking changes in nunit
- temporarily removed `readonly` from members of `ApplicationParameters` which need to be mutable for the integration tests _(the old dirty workaround using reflection to change readonly fields does not work in .net 5)_
- replaced custom `HashCode` implementation with `System.HashCode`
- removed usage of Cng HashAlogrithms
- removed `SecurityProtocolType` workaround

Known issues and TODOs:
- the Uppercut build scripts don't work and no newer version is available
- I didn't touch the `Rhino.Licensing` lib. Should be upgraded to .net 5 or standard as well.
- I'm certain there are many more custom implementations and workarounds removable and more optimizations possible. _(E.g. FluentAssertions offers many specialized functions that could replace whole constructs in the tests)_
- There seems to be a breaking change in the file locking which causes the three `should_not_be_able_delete_the_rollback` test to fail

#### **nuget-chocolatey**
For now I only upgraded the `Core`, `Test.Core` and `Test.Utility` projects and unloaded the rest of the projects to build the library I used in `choco`.
All unit tests succeed.

Here is a list of the changes I made _(omitting some smaller adjustment details)_:
- upgraded projects to .net 5
- upgraded dependencies to latest versions
- replaced local copy of `xdt` with latest upstream nuget package
- adjusted to breaking changes of `xunit`
- fixed or ignored various new diagnostic warnings
- added `Microsoft.Data.Services.Client` as replacement for stuff missing .net 5
- temporarily removed unused class `AssemblyMetadataExtractor` due to compatibility issues

Known issues and TODOs:
- I'm not aware which components are actually used for other parts of the ecosystem, so there might be the need to upgrade other parts as well.
- I'm certain there are many more custom implementations and workarounds removable and more optimizations possible.

---

### Conclusion
In my opinion it's absolutely feasible and worth to upgrade to .net 5.
Granted I don't know the all the finer details about the ecosystem and the reasons for sticking with .net framework 4, but that's why I opened this issue to get some feedback before I continue my efforts.

If there are reasons for sticking with .net framework 4 or you simply don't want to upgrade at all, feel free to close the issue; no hard feelings.
For me this was just a nice little exercise in .net migrations and the time was definitely not wasted, because I learned quite a bit doing it and it didn't take long at all.

There is definitely a bunch more work that would need to be completed for an actual upgrade since it would affect other parts of the ecosystem like chocolatey GUI, but I would be willing to invest more time into it if there is a reasonable chance of getting it merged at some point.

In case such a migration effort would be something you want to do, I would appreciate any pointers you can give me on missing parts, improvements and any other suggestion on what to tackle next.

**Comment by TheCakeIsNaOH** [comment]
> replaced local version of System.Management.Automation with latest upstream nuget package

I think the source code tree has an older version of this for a specific reason. I do know it is specifically not ilmerge-d into the final `choco.exe`, from what I understand, so the system version of that assembly can be used.

> the Uppercut build scripts don't work and no newer version is available

Open issue here to replace them:
https://github.com/chocolatey/choco/issues/2400

> The goal of this issue is to gauge if there is even an interest in upgrading and to gather feedback before committing more time or opening a PR.

Upgrading the .Net version is already planned:
https://github.com/chocolatey/choco/issues/2173

But from what I understand, getting the nuget library updated to a newer base version is the next big TODO item for Chocolatey CLI. Could change of course.
https://github.com/chocolatey/choco/issues/508

Personally, I'm really excited for the day that choco can run on .Net core/5/6. Since Mono is starting to annoy me. I've found a couple of bugs in mono, and the debugging support sucks compared to "real" .Net.

**Comment by jan-guenter** [comment]
@TheCakeIsNaOH thanks for your quick reply.

> > replaced local version of System.Management.Automation with latest upstream nuget package
> 
> I think the source code tree has an older version of this for a specific reason. I do know it is specifically not ilmerge-d into the final `choco.exe`, from what I understand, so the system version of that assembly can be used.

Do you know where I would find more details on that topic or d o you know someone who knows?
Would love to learn the reason for it. after adding the Powershell SDK I was actually able to remove the whole assembly resolution and loading magic and it seems to work fine with my system powershell. I tried it with the latest powershell 7 version installed as well as after uninstalling it, leaving my Windows 10 system with its default version 5.

> > the Uppercut build scripts don't work and no newer version is available
> 
> Open issue here to replace them: #2400

I was wondering if it could simply be replace by msbuild tasks. They have become very flexible and we could code custom extensions if necessary. Need to take a closer look and will comment in that issue.

> > The goal of this issue is to gauge if there is even an interest in upgrading and to gather feedback before committing more time or opening a PR.
>
> Upgrading the .Net version is already planned: #2173

I don't know why I didn't find the upgrade ticket; thanks for referencing it! Sometimes the github search is not as helpful as one  might think :-D
 
> But from what I understand, getting the nuget library updated to a newer base version is the next big TODO item for Chocolatey CLI. Could change of course. #508

That sound reasonable. I wasn't aware of all the customization performed to it, so I just ported the necessary parts of the v2 version.

> Personally, I'm really excited for the day that choco can run on .Net core/5/6. Since Mono is starting to annoy me. I've found a couple of bugs in mono, and the debugging support sucks compared to "real" .Net.

I feel your pain! In my professional life I also had to deal with mono in the past and also worked on some new cross-platform .net 5 projects.

**Comment by TheCakeIsNaOH** [comment]
> Do you know where I would find more details on that topic or d o you know someone who knows?

CC @gep13 

> I was wondering if it could simply be replace by msbuild tasks. They have become very flexible and we could code custom extensions if necessary. 

I suspect that cake will be the replacement, since that is what ChocolateyGUI, the Chocolatey docs repo, and probably the closed source extensions all use. Also, multiple people around here contribute to it, or to extensions.

**Comment by gep13** [comment]
>@jan-guenter said...
I was surprised to find that chocolatey still uses .net Framework 4, so I took a quick look on how complicated an upgrade to .net 5 would be and created a proof of concept.

Chocolatey has a long standing tradition of supporting as many operating systems, PowerShell versions, and .Net Framework versions as possible.  As a result, we can say that it is possible to run Chocolatey on an operating system that uses PowerShell 2 and .NET Framework 4 (for example Windows Server 2008 R2).  This might seem surprising to you, but there are a number of folks that are still in that position, and for the time-being, running Chocolatey in these environments is still going to be an option.

Having said that, bringing Chocolatey up to run on the latest frameworks is something that we have spoken about (including recently) but this is not something that we can jump on immediately, so I want to set some expectations here.  This is not something that will happen for months at the earliest.

In order to get to a point where we could make a fundamental change like this, we need to give folks time to prepare, and we also need to signal this in the version numbers that we have.  As a rough plan, I would see it going something like this (**NOTE:** This is far from decided, and will be subject to change):

* Release a 1.0.0 version of Chocolatey (as well as major releases of all Chocolatey components, i.e. Chocolatey GUI, Chocolatey Agent, Chocolatey Licensed, etc).
  * This would continue to use the current target .NET Framework versions
* Release a 2.0.0 version of Chocolatey (as well as major releases of all Chocolatey components, i.e. Chocolatey GUI, Chocolatey Agent, Chocolatey Licensed, etc).
  * This would bump the required .NET Framework version to .NET 4.8
  * This would switch to new csproj format, and PackageReferences
  * This would update all dependencies, removing what is no longer required
  * Include new build infrastructure, likely Cake
* Release a 3.0.0 version of Chocolatey (as well as major releases of all Chocolatey components, i.e. Chocolatey GUI, Chocolatey Agent, Chocolatey Licensed, etc).
  * This would switch to using .NET 5 (or more likely .NET 6)
  * This would attempt to use pre-compiled, standalone executable (ready to run) that don't take a dependency on an installed version of .NET Runtime/SDK
  * This would likely include an update to the NuGet assemblies to support v3 feeds

> @jan-guenter said...
I didn't touch the Rhino.Licensing lib. Should be upgraded to .net 5 or standard as well.

Rhino.Licensing has already been updated to target multiple frameworks.  You can see this here:

https://github.com/chocolatey/rhino-licensing/blob/master/Rhino.Licensing/Rhino.Licensing.csproj#L5

> @jan-guenter said...
Granted I don't know the all the finer details about the ecosystem and the reasons for sticking with .net framework 4, but that's why I opened this issue to get some feedback before I continue my efforts.

There are lots of additional Chocolatey components/products.  So any change to something as fundamental as this is going to take time.  Updating Chocolatey CLI would then have knock on updates to Chocolatey Licensed, Chocolatey Agent, Chocolatey GUI, Chocolatey GUI Extension, Boxstarter, etc.

> @jan-guenter said...
If there are reasons for sticking with .net framework 4 or you simply don't want to upgrade at all, feel free to close the issue; no hard feelings.

This is definitely something that we _want_ to do, but again, I want to be honest in terms of setting the stage with regard to _when_ this would happen.

> @TheCakeIsNaOH said...
I think the source code tree has an older version of this for a specific reason. I do know it is specifically not ilmerge-d into the final choco.exe, from what I understand, so the system version of that assembly can be used.

Yes, exactly, this assembly isn't ILMerged into the choco.exe so that the system version of SMA is used.  Again, this is for historical reasons, so that we can continue to support PowerShell 2.

> @jan-guenter said...
I was wondering if it could simply be replace by msbuild tasks. They have become very flexible and we could code custom extensions if necessary. Need to take a closer look and will comment in that issue.

It has pretty much been decided that we will be switching to Cake.  As @TheCakeIsNaOH mentioned, Cake is already being used in other places within the Chocolatey eco-system, so we will be looking to keep that consistency here also.

**Comment by jan-guenter** [comment]
@gep13 Thank you very much for this detailed response!

I totally get and respect it if you want to keep support for older systems.
I have been working as a software developer for over a decade in the banking sector and I know that there are still many outdated systems out there. Most often businesses simply don't want to invest the money for new hardware and are therefore stuck with older operating systems.

I never expected such a migration to be done anytime soon, but I couldn't find any communication that it is planned at all. That's why I thought a POC might help to sway you in favor of a migration, but most importantly it was good exercise for myself and well worth the time even if you don't want to upgrade at all.

The rough roadmap you proposed for the migration seems reasonable, but in that scenario I would suggest to do the first migration step to 4.6 not 4.8, so that version 2 would still be working on Vista.

Cake sound like a very good fit. I've seen it before, but never used it myself. Since @TheCakeIsNaOH mentioned it I studied the docs and I'm impressed by its functionality and structure.

If you like you can close the issue, since I assume you will create some kind of Epic for this anyway when you made your decisions on how to proceed. Or you can leave it open as a reference; perhaps you want to use some of the stuff I did for the POC; that's totally up to you guys.

**Comment by gep13** [comment]
@jan-guenter I will leave this issue open for now, as there is some very useful information here.  It may be that it will eventually get closed, and information moved elsewhere.

**Comment by gep13** [comment]
Can I ask that you keep your branches around, as these could be a useful source of information when we come to this work?

**Comment by jan-guenter** [comment]
> Can I ask that you keep your branches around, as these could be a useful source of information when we come to this work?

No problem and feel free to tag me in other topics if there are any questions.

**Comment by tapika** [comment]
Ive ported to net 4.8, .net core 3.1, net 5.0 and 6.0.

https://github.com/tapika/swupd

That should include choco + nuget core + choco gui + some of microsoft libs.

Would be happy to exchange patches.

**Comment by tapika** [comment]
But havent ported tests yet, that is something i need to check next.

**Comment by jan-guenter** [comment]
@tapika feel free to cherry pick from my forks.
I'm not a man of million commits, so my whole upgrade process was done in 3 commits since it only was a weekend project. I tried to touch the existing code as little as possible, but the test were important to me to verify my work, so I corrected the codebase only to the degree necessary to achieve a working version.

I'll definitely take a look at your branch the next days, because I'm curious how you solved the `IWeakEventListener` dependency. I pulled in the whole WPF framework just for that one; there must be a better solution.

**Comment by tapika** [comment]
@jan-guenter May I ask - why do you need this chocolatey upgrade - do you have any further plans for it ?

**Comment by tapika** [comment]
> Chocolatey has a long standing tradition of supporting as many operating systems, PowerShell versions, and .Net Framework versions as possible. As a result, we can say that it is possible to run Chocolatey on an operating system that uses PowerShell 2 and .NET Framework 4 (for example Windows Server 2008 R2). This might seem surprising to you, but there are a number of folks that are still in that position, and for the time-being, running Chocolatey in these environments is still going to be an option.

This is important for me as well, but I suspect that this could be achieved via ReadyToRun - however - I haven't tested yet in virtual machines. My plan however is - if this will be a problem - I will switch from powershell to plain executable starting.
But this is something I need to test still. 

What I've realized - is while I'm performing upgrade - installer update changes as well, like dropping windows 7 support and so on.

**Comment by jan-guenter** [comment]
> @jan-guenter May I ask - why do you need this chocolatey upgrade - do you have any further plans for it ?

@tapika see my initial post section "Background":
> I planned to improve a little bit on the chocolatey GUI and was surprised to see the project using .net Framework 4, so I decided to start from scratch using .net 5 and WPF to create a more native feeling UI.
> When trying to add the chocolatey lib I realized that it also is only available for .net Framework.
> So I decided to have look how complicated it would be to upgrade choco.
> To my surprise the upgrade was rather painless and made a POC in my fork.

**Comment by tapika** [comment]
Btw, concerning unit tests. I've briefly checked your changes in unit tests, looks nice - however - 
I would like to have code coverage enabled. (My git contains coverage badge icons with 50% code covered at the moment.)

https://www.lambdatest.com/blog/nunit-vs-xunit-vs-mstest/
https://stackoverflow.com/questions/9769047/nunit-vs-xunit

There are some opinions on which test framework to choose, but
probably as long as test framework satisfies your needs - then go for it.
(some thinks like this).

I have been using xunit v2 by myself and even made some patches to it -
but still - I'm waiting for xunit v3 nuget packages to become available.

Now I find out that xunit v3 did came out as an alpha:
https://xunit.net/docs/v3-alpha

But what I've briefly tested - it's missing visualstudio.xunit adapter,
which is apparently separate git: https://github.com/xunit/visualstudio.xunit
and not supported by v3. hmm...

Then I've started to think if I can support code coverage with xunit and found an article about it:
http://derekwilson.net/blog/2012/05/29/using-opencover-and-xunit

So I think I'll try out xunit v2 + opencover.

But I would prefer not to convert all tests immediately, but as I add new tests instead.

Basically I could leave chocolatey.tests, chocolatey.tests.integration as they are
and create new unit tests instead and maybe start to move test by test into new test 
framework.

But need to check if code coverage statistics can be somehow merged.

If you have ideas / proposals which way to go, let me know.

**Comment by jan-guenter** [comment]
@tapika
What are trying to achieve here? This issue was about an upgrade of the .net version. 
All other changes where necessary, unavoidable consequences. I only thrown out `TinySpec` and replaced the `Should` library, which was pulled in by `TinySpec`, with `FluentAssertions` due to the fact that `Should` and `TinySpec` are no longer maintained.
I chose `FluentAssertions`, because it is well know, has good support and a very similar fluent syntax as `Should`.
For `TinySpec` I also just removed the package reference since it does not contain a library but only adds the `TinySpec.cs` helper code which I left in place. 

Such a migration is not a trivial task and should be done in well defined steps with minimal impact each step. Yes, switching to a different testing framework could be such a step in a later stage, but that would need some agreement of the team first.
To quote the [Contribution Guidelines - Contributing Process](https://github.com/chocolatey/choco/blob/develop/CONTRIBUTING.md#contributing-process):
> Once you get a nod from one of the Chocolatey Team, you can start on the feature.

Replacing the whole testing framework would be a completely different discussion.
I would like to ask you to stay on topic here and open separate discussions for unrelated topics like that.

That said, it would be rude of me not to respond to your request for ideas and proposals, since you already took the time to write this up.
So here are my two cents on that topic:
- You can achieve any test scenario with every test framework.
  It basically boils down to personal preference and tooling support.
- Having multiple test frameworks in the same solution just leads to confusion and I would advise against it.
- Writing good quality tests is way more important than test frameworks.
- Microsoft maintains a very nice section in the MSDN about testing approaches for .net.
  This entry is constantly updated and covers xUnit, NUnit and MSTest as well as code coverage and general best practices.
  https://docs.microsoft.com/en-us/dotnet/core/testing/

**Comment by tapika** [comment]
> @tapika What are trying to achieve here? 

Ok, maybe I'm saying this bit late, but I'm not a part of chocolatey team, and not working on main chocolatey fork
(https://github.com/chocolatey/choco) , but on my own (https://github.com/tapika/swupd) - and because of that one, I want to have different requirements than original chocolate team.

Code coverage badge:
![image](https://user-images.githubusercontent.com/3228645/140606806-d16e7b4c-1b10-4cc2-8460-256daaa27f88.png)

Exists only in my fork, not in official chocolatey.

> Such a migration is not a trivial task and should be done in well defined steps with minimal impact each step. Yes, switching to a different testing framework could be such a step in a later stage, but that would need some agreement of the team first. To quote the [Contribution Guidelines - Contributing Process](https://github.com/chocolatey/choco/blob/develop/CONTRIBUTING.md#contributing-process):
> 
> > Once you get a nod from one of the Chocolatey Team, you can start on the feature.

This is applicable to Chocolatey team, so they can accept / do not accept changes.
Once I have started .net upgrade - I've realized that I want to recombine multiple git repos into one, but
this change was never approved by Chocolatey team - that's why I don't create any pull request to chocolatey.

At the moment me and Chocolatey team only checking what other side is doing and maybe just pick up some changes
which we like from each other.

Moverover - one change related to integrity check 

SHA-1: 7871df46257a4582a7248fdfa62efba5bfd862cd

* bugfix: ProtectedData.Protect / Unprotect works only on Windows machines (See `https://github.com/dotnet/runtime/issues/22886`)

most probably will never be accepted by Chocolatey team - they did not like this fix at least.

> Replacing the whole testing framework would be a completely different discussion. I would like to ask you to stay on topic here and open separate discussions for unrelated topics like that.

I will analyze this further, I haven't yet told that I will replace it. 
And I'm now talking only about my own fork (https://github.com/tapika/swupd) - what it will have at the end.

Your changes still might get accepted into official chocolatey fork. (I'm not guy who makes that decision)

But from future chocolatey development perspective, maybe when we will merge all forks into one, what kind of testing framework you want to have at the end. 

What I'm asking you - can you make a wish, and what it would be ? 

> * You can achieve any test scenario with every test framework.
>   It basically boils down to personal preference and tooling support.
> * Having multiple test frameworks in the same solution just leads to confusion and I would advise against it.
> * Writing good quality tests is way more important than test frameworks.
> * Microsoft maintains a very nice section in the MSDN about testing approaches for .net.
>   This entry is constantly updated and covers xUnit, NUnit and MSTest as well as code coverage and general best practices.
>   https://docs.microsoft.com/en-us/dotnet/core/testing/

Agree on all statements above. I'm at the moment testing how testing will work on .net core 3.1, then will also try code coverage, and then will try also other platforms (.net 5.0, 6.0, .net 4.8).

At the moment my plan is to keep all platforms working, until I have clear idea why some of platform does not need to be supported. (Mapped to ReadyToRun / virtual machine testing).

Again - what I'm testing is different what Chocolatey team is testing.
Chocolatey official fork uses custom `ilmerge` tool, not `ReadyToRun`. (Dont know all details).

**Comment by jan-guenter** [comment]
> Ok, maybe I'm saying this bit late, but I'm not a part of chocolatey team, and not working on main chocolatey fork
> (https://github.com/chocolatey/choco) , but on my own (https://github.com/tapika/swupd) - and because of that one, I want to have different requirements than original chocolate team.

The part of you not being one of the choco team members was clear, but I wrongly assumed that we have the same goal of contributing to the upstream version.

It is of cause absolutely up to you what you do in your fork if you don't plan to get it merged anyways.

However my point regarding this discussion being off-topic for this issue thread still stands. Having information regarding a certain issue/request cluttered with off-topic conversations like this is, in my opinion, not helpful. There are other ways and platforms for that.
If you want to talk to me explicitly simply write an email or if you like to have more input from others of the choco community I would suggest you try the [choco discord server](https://ch0.co/community). You can also find me on the discord server with user id jellow#9966.

**Comment by tapika** [comment]
> If you want to talk to me explicitly simply write an email or if you like to have more input from others of the choco community I would suggest you try the [choco discord server](https://ch0.co/community). You can also find me on the discord server with user id jellow#9966.

Agree, lets continue over mail / discord. ( I did send you e-mail with same idea ).

**Comment by gep13** [comment]
@jan-guenter I wanted to follow up with this issue, as we have been making some decisions about the future direction of Chocolatey.  The information that you have provided here is very useful, and will come in handy when we make the jump to .NET (Core).  In the interim, in the next release of Chocolatey, we are going to bump to .NET Framework 4.8, which takes us along the path to .NET (Core).

The issue for tracking this work is here: https://github.com/chocolatey/choco/pull/2739

For now, I am going to close this issue, but please keep in mind, that getting to .NET (Core) is still the ultimate goal, we just need to work our way there in steps, and getting to .NET Framework 4.8, is the first of those.

**Comment by tapika** [comment]
FYI @gep13 @TheCakeIsNaOH @jan-guenter 

Please notice that I have already upgraded in my own [fork](https://github.com/tapika/swupd) to .net framework 4.8 (cli + gui), as well as other nets - like .net core 3.1 (cli + gui), .net 5.0 (only cli), .net 6.0 (only cli). Currently using .net 3.1 core as a base platform I'm working on, but checking other .net's from time to time. So not maintaining actively side .net's, but sometimes checking them through so they would still work.

Feel free to pick up same .sln folder structure as I have in [here](https://github.com/tapika/swupd/tree/master/src).
